File: laby/src/main/java/agh/ics/oop/MapDirections.java
package agh.ics.oop;

public enum MapDirections {
    NORTH,
    SOUTH,
    WEST,
    EAST,
    NONE;

    @Override
    public String toString() {
        return switch (this){
            case NORTH -> "Północ";
            case SOUTH -> "Południe";
            case WEST -> "Zachód";
            case EAST -> "Wschód";
            default -> "Brak kieunku";
        };
    }

    public MapDirections next(){
        return switch (this){
            case NORTH -> EAST;
            case SOUTH -> WEST;
            case WEST -> NORTH;
            case EAST -> SOUTH;
            default -> NONE;
        };
    }

    public MapDirections previous(){
        return switch (this){
            case NORTH -> WEST;
            case SOUTH -> EAST;
            case WEST -> SOUTH;
            case EAST -> NORTH;
            default -> NONE;
        };
    }
    public Vector2d toUnitVector(){
        return switch (this){
            case NORTH -> new Vector2d(0,1);
            case SOUTH -> new Vector2d(0,-1);
            case WEST -> new Vector2d(-1,0);
            case EAST -> new Vector2d(1,0);
            default -> new Vector2d(0,0);
        };
    }

}

File: KotlinLab/src/main/kotlin/MapDirections.kt
package ooplab

enum class MapDirections {
    NORTH,
    SOUTH,
    WEST,
    EAST,
    NONE;

    override fun toString(): String {
        return when (this) {
            NORTH -> "Północ"
            SOUTH -> "Południe"
            WEST -> "Zachód"
            EAST -> "Wschód"
            else -> "Brak kieunku"
        }
    }

    operator fun next(): MapDirections {
        return when (this) {
            NORTH -> EAST
            SOUTH -> WEST
            WEST -> NORTH
            EAST -> SOUTH
            else -> NONE
        }
    }

    fun previous(): MapDirections {
        return when (this) {
            NORTH -> WEST
            SOUTH -> EAST
            WEST -> SOUTH
            EAST -> NORTH
            else -> NONE
        }
    }

    fun toUnitVector(): Vector2d {
        return when (this) {
            NORTH -> Vector2d(0, 1)
            SOUTH -> Vector2d(0, -1)
            WEST -> Vector2d(-1, 0)
            EAST -> Vector2d(1, 0)
            else -> Vector2d(0, 0)
        }
    }
}